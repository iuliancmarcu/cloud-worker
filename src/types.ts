export type TaskType = 'FIBO' | 'FACT';

export type Task = {
  id: string;
  type: TaskType;
  payload: object;
};

export type CompletedTask = Task & {
  result: object;
};

export type FiboTask = Task & {
  type: 'FIBO';
  payload: {
    n: number;
  };
};

export type FactTask = Task & {
  type: 'FACT';
  payload: {
    n: number;
  };
};
