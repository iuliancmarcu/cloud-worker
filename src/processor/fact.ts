import { CompletedTask, FactTask } from '../types';

export async function process(task: FactTask): Promise<CompletedTask> {
  const fact = factorial(task.payload.n);

  return {
    ...task,
    result: {
      value: fact,
    },
  };
}

function factorial(n: number): number {
  if (n <= 1) {
    return 1;
  }

  return n * factorial(n - 1);
}
