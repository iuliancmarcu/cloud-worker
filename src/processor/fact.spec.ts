import * as fact from './fact';
import { FactTask } from '../types';

describe('Processor: fact', () => {
  it('processes a FACT task and resolves with a CompletedTask', () => {
    const task: FactTask = {
      id: 'random-id',
      type: 'FACT',
      payload: {
        n: 10,
      },
    };

    expect(fact.process(task)).resolves.toMatchSnapshot();
  });
});
