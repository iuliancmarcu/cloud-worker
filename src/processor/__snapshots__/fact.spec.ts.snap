// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processor: fact processes a FACT task and resolves with a CompletedTask 1`] = `
Object {
  "id": "random-id",
  "payload": Object {
    "n": 10,
  },
  "result": Object {
    "value": 3628800,
  },
  "type": "FACT",
}
`;
